state/machine.go

```diff
@@ -2237,40 +2239,56 @@ func (m *Machine) SetUpgradeCharmProfile(appName, chURL string) error {
 				return nil, errors.Trace(err)
 			}
 		}
+
 		var emptyProfile bool
 		if ch == nil || (ch.LXDProfile() == nil || ch.LXDProfile().Empty()) {
 			emptyProfile = true
 		}
 
-		provisioned := true
+		ops := []txn.Op{
+			{
+				C:      machinesC,
+				Id:     m.doc.DocID,
+				Assert: isAliveDoc,
+			},
+			// TODO: lxd profile - enhance checks for this set of transactions
+			//}, {
+			//	C:      charmsC,
+			//	Id:     ch.doc.DocID,
+			//	Assert: bson.D{{"url", ch.URL()}},
+			//},
+		}
+
 		profiles, err := m.CharmProfiles()
 		if err != nil {
 			if !errors.IsNotProvisioned(err) {
 				return nil, errors.Trace(err)
 			}
-			provisioned = false
-		}
-		// we need to ensure that the machine is provisioned before we attempt
-		// to set that the upgrade charm profile complete to not required.
-		// If the profiles are empty and there is nothing to apply, we can
-		// assume that nothing is required.
-		if len(profiles) == 0 && emptyProfile {
-			var ops []txn.Op
-			if provisioned {
-				ops = append(ops, m.checkCharmProfilesIsEmptyOp())
+		} else {
+			ops = append(ops, txn.Op{
+				C:      instanceDataC,
+				Id:     m.doc.DocID,
+				Assert: bson.D{{"charm-profiles", profiles}},
```

Comment:
I should clarify that I think this mechanism of clearly asserting the list of profiles hasn't changed since we looked at it is correct. It doesn't matter if the old value is an empty list or one with things in it.
The important bit is that we're making assumptions of what to do based on the list of profiles that we read. If that list changes concurrently, then we need to reevaluate what to do.

Reply:
There is a case where this isn't correct.  "charm-profiles" is setup with omit empty, so if there are no charm-profiles, this check fails.  Babbageclunk helped me find it. Will be updating this code shortly.